text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,487 @@
+
+![Pipelined](images/otlp-pipelined.png)
+
+The number of unacknowledged requests for which the client did not receive responses yet is defined by the window size of underlying transport (for gRPC see details about [how the window size is calculated](https://grpc.io/blog/2017-08-22-grpc-go-perf-improvements/)).
+
+If the client detects that the underlying transport is broken and must be re-established the client will optionally retry sending all requests that are still pending acknowledgement after re-establishing the transport. The client implementation SHOULD expose an option to turn on and off the retrying; by default retrying SHOULD be on.
+
+If the client is shutting down (e.g. when the containing process wants to exit) the client will optionally wait until all pending acknowledgements are received or until an implementation specific timeout expires. This ensures reliable delivery of telemetry data. The client implementation SHOULD expose an option to turn on and off the waiting during shutdown.
+
+If the client is unable to deliver a certain request (e.g. it retried but failed multiple times or shutdown timer expired while waiting for acknowledgements) the client SHOULD record the fact that the data was not delivered.
+
+Pipelined mode is recommended when the client and the server are connected via a high-latency network but high throughput is still required. This is a typical situation when the client is an application running in one data center and the server is a telemetry backend running in a different data center, potentially in a completely different geography that is reachable via high-latency multi-hop network. Same recommendation applies to low latency networks if the server is capable of high processing rate but takes a long time to generate a response to each particular request.
+
+Note that the server does not need to know in what mode the client operates. The server just needs to ensure that it sends a response to every request. The Synchronized or Pipelined mode is enabled purely by client behavior.
+
+### Server Response
+
+The Server replies to every `Export` request received from the client by a corresponding `Export` response. The ID field of the response is set equal to the ID field of the request.
+
+If the Client does not receive a response to a particular request after a configurable period of time then the client SHOULD drop the request. The client SHOULD maintain a counter of such dropped data. This is the recommended approach because the reason that the response is not received may be that the server fails to generate a response to this particular request. Attempting to re-try sending may cause more problems at the server side. This timeout SHOULD be configured to be significantly larger than it normally takes the server to respond.
```

Comment:
Implementing Deadline on top of a bidirectional stream may be a hard and error prone thing. I need to think a bit more about this.